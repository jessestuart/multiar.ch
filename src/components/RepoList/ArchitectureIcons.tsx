import { Architecture } from 'docker-hub-utils'
import React from 'react'
import styled from 'styled-components'

import { colors, fonts } from '../../styles/Theme'

export const ArchitectureIcon = styled.span`
  background-color: ${colors.secondary};
  border-radius: 10%;
  border: 1px solid #abb7b2;
  display: inline;
  font-family: ${fonts.mono};
  user-select: none !important;
`

// eslint-disable-next-line
export const ArchitectureIconGroup = ({ architectures }) => (
  <>
    {[...architectures].sort().map((architecture: Architecture) => (
      <ArchitectureIcon key={architecture} className="f7 pa1 ma1">
        {architecture}
      </ArchitectureIcon>
    ))}
  </>
)
