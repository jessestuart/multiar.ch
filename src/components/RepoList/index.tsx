import { graphql, useStaticQuery } from 'gatsby'
import { DockerHubRepo } from 'gatsby-source-docker-hub'
import _ from 'lodash'
import fp from 'lodash/fp'
import { DateTime } from 'luxon'
import React from 'react'
import { Box, Text } from 'rebass'

import Theme from 'styles/Theme'

const RepoList = () => {
  const query = useStaticQuery(graphql`
    {
      allDockerHubRepo {
        edges {
          node {
            description
            id
            lastUpdated
            name
            pullCount
            starCount
          }
        }
      }
    }
  `)

  const repos: DockerHubRepo[] = _.flow(
    fp.get('allDockerHubRepo.edges'),
    fp.map('node'),
    // fp.filter(repo => DateTime.fromISO(repo.lastUpdated).diffNow()),
  )(query)
  if (_.isEmpty(repos)) {
    return null
  }

  return (
    <Box className="justify-center center flex pt4 f4">
      <ol className="mw7-ns w-100">
        <li className="flex fw7 lh-copy">
          <Text className="flex-nowrap w-100" fontFamily="mono">
            Image
          </Text>
          <Text className="flex justify-end w-100" fontFamily="mono">
            Pulls
          </Text>
        </li>
        {repos.map((repo: DockerHubRepo, index: number) => {
          const lastUpdatedRelative = DateTime.fromISO(
            _.toString(repo.lastUpdated),
          ).toRelative()

          return (
            <li key={index} className="flex lh-copy pt2 pb3 bb b--near-white">
              <div className="w-100">
                <a
                  className="flex-nowrap primary lh-title"
                  style={{ color: Theme.colors.primary }}
                  href={`https://cloud.docker.com/u/jessestuart/repository/docker/jessestuart/${repo.name}`}
                >
                  <Text fontFamily="mono" color="primary" className="fw5">
                    {repo.name}
                  </Text>
                </a>
                <Text>{repo.description}</Text>
                <Text
                  className="flex justify-start no-underline flex-column f6"
                  fontFamily="sans-serif"
                  style={{ color: Theme.colors.textMuted }}
                >
                  Updated {lastUpdatedRelative}
                </Text>
              </div>
              <Text
                className="pl4"
                style={{ display: 'inline-block' }}
                fontFamily="mono"
              >
                {repo.pullCount.toLocaleString()}
              </Text>
            </li>
          )
        })}
      </ol>
    </Box>
  )
}

export default RepoList
