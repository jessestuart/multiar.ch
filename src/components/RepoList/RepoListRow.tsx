import classNames from 'classnames'
import { Architecture, DockerHubRepo } from 'docker-hub-utils'
import _ from 'lodash'
import { DateTime } from 'luxon'
import React from 'react'
import { Link, Text } from 'rebass/styled-components'
import { Box } from 'reflexbox'

import { ArchitectureIconGroup } from 'components/RepoList/ArchitectureIcons'

const DOCKER_HUB_URL = 'https://hub.docker.com/r/jessestuart/'

interface Props {
  className?: string
  repo: DockerHubRepo
  architectures: Architecture[]
  style?: any
}

const getLastUpdatedRelative = (repo: DockerHubRepo): string | null =>
  DateTime.fromISO(_.toString(repo.lastUpdated)).toRelative()

const RepoListRow = (props: Props) => {
  const { architectures = [], className, repo } = props
  const lastUpdatedRelative: string | null = getLastUpdatedRelative(repo)
  const repoUrl = `${DOCKER_HUB_URL}${repo.name}`

  return (
    <li
      key={repo.name}
      className={classNames('flex lh-copy pt2 pb3', className)}
      style={props.style}
    >
      <Box>
        <Link
          className="flex-nowrap inline-flex primary"
          color="primary"
          href={repoUrl}
        >
          <Text fontFamily="mono" className="fw5 pr2">
            {repo.name}
          </Text>
        </Link>
        <ArchitectureIconGroup architectures={architectures} />
        <Text className="lh-title pv2">{repo.description}</Text>
        {lastUpdatedRelative ? (
          <Text
            className="flex flex-column f6 justify-start no-underline"
            fontFamily="sans"
            color="textMuted"
          >
            Updated {lastUpdatedRelative}
          </Text>
        ) : null}
      </Box>
      <Text className="flex flex-auto justify-end pl4" fontFamily="mono">
        {repo.pullCount.toLocaleString()}
      </Text>
    </li>
  )
}

export default RepoListRow
