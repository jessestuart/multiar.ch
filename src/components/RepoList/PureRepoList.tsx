import { DockerHubRepo } from 'docker-hub-utils'
import { Architecture } from 'docker-hub-utils'
import _ from 'lodash'
import React from 'react'
import { Flex, Text } from 'rebass/styled-components'

import { PullCountMap } from 'components/RepoList'
import RepoListRow from 'components/RepoList/RepoListRow'
import { colors } from 'styles/Theme'

interface Props {
  initialRepoPullCount?: PullCountMap
  repos: DockerHubRepo[] | undefined
  repoToArchitecturesMap: { [repoName: string]: Architecture[] }
}

const PureRepoList = ({
  initialRepoPullCount,
  repos,
  repoToArchitecturesMap,
}: Props) => {
  if (!repos || _.isEmpty(repos)) {
    return null
  }

  return (
    <Flex className="center f4 justify-center pt4">
      <ol className="mw7-ns">
        <li className="flex fw7 lh-copy">
          <Text className="flex-nowrap flex-auto" fontFamily="mono">
            Image
          </Text>
          <Text className="flex justify-end" fontFamily="mono">
            Pulls
          </Text>
        </li>
        {_.sortBy(repos, 'pullCount')
          .reverse()
          .map((repo: DockerHubRepo, index: number) => {
            return (
              <RepoListRow
                key={repo.name}
                initialRepoPullCount={initialRepoPullCount}
                repo={repo}
                architectures={repoToArchitecturesMap[repo.name]}
                style={
                  index === repos.length - 1
                    ? null
                    : { borderBottom: `1px ${colors.neutral} solid` }
                }
              />
            )
          })}
      </ol>
    </Flex>
  )
}

export default PureRepoList
