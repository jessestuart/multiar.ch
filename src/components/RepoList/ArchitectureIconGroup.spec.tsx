import React from 'react'
import renderer from 'react-test-renderer'

import { ArchitectureIconGroup } from 'components/RepoList/RepoListRow'

describe('ArchitectureIconGroup', () => {
  test('render with default icons', () => {
    const component = renderer.create(
      <ArchitectureIconGroup architectures={['amd64', 'arm64', 'arm']} />,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
