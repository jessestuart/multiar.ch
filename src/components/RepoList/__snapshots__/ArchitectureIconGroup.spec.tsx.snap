// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchitectureIconGroup render with default icons 1`] = `
Array [
  <span
    className="f7 pa1 ma1 sc-bdVaJa kJZcve"
  >
    amd64
  </span>,
  <span
    className="f7 pa1 ma1 sc-bdVaJa kJZcve"
  >
    arm
  </span>,
  <span
    className="f7 pa1 ma1 sc-bdVaJa kJZcve"
  >
    arm64
  </span>,
]
`;

exports[`ArchitectureIconGroup render with no icons 1`] = `null`;
